### PR TITLE
Improve cache invalidation for bar and hud segment

### PR DESCRIPTION
**Pull request by marienz**
This makes the changes described in comments on #439.

I tested the cases mentioned there, but this could do with more testing before
being merged.

### CODE REVIEW

**Review by seagle0128 — APPROVED** [review]

File: doom-modeline-segments.el [hunk 1]
```diff
@@ -1359,36 +1359,38 @@ of active `multiple-cursors'."
 (doom-modeline-def-segment bar
   "The bar regulates the height of the mode-line in GUI."
   (when doom-modeline-bar
+    (unless (and doom-modeline--bar-active doom-modeline--bar-inactive)
+      (let ((width doom-modeline-bar-width)
+            (height (max doom-modeline-height
+                         (doom-modeline--font-height))))
+        (when (and (numberp width) (numberp height))
+          (setq doom-modeline--bar-active
+                (doom-modeline--make-image 'doom-modeline-bar width height)
```
Comment (marienz) [inline doom-modeline-segments.el:1368]: This doesn't look right: it's undoing the rename of doom-modeline--make-image to doom-modeline--create-bar-image (looks like something went wrong merging master). I've sent #444 to fix it.

Comment (seagle0128) [reply doom-modeline-segments.el:1368]: It's a merge issue. I've already fixed it. 3b39376

**Comment by seagle0128** [comment]
@marienz I got this error while I was using `company-box`. Any idea?

```elisp
Error during redisplay: (eval (doom-modeline-segment--hud)) signaled (wrong-type-argument wholenump -34228)
```

**Comment by marienz** [comment]
Interesting. I did get that before but I thought the `(max 0 ...)` for `bottom-margin` fixed it: it happens when one of the `make-string` calls in `doom-modeline--create-hud-image` gets called with a negative length (so if one of the margins is negative, or if the margins added together are greater than the height). I didn't spot any other ways of triggering this.

I tried enabling `company-box` and triggering completion, but that doesn't seem to be sufficient to reproduce it. Can you reproduce this (or do you have an idea of what might be triggering it if not just company-box)?

Looking at the calculations again:

- `top-margin` can't be negative, because `doom-modeline-height`, `(buffer-size)` and `(1- (window-start))` can't be negative (positions are at least 1, so subtracting 1 is safe).
- `bottom-margin` can't be negative because of that `(max 0 ...)`.
- If shrinking the buffer a lot updates `(buffer-size)` but does not immediately update `(window-start)` or `(window-end)`, that might cause this. We would clamp `bottom-margin` to 0, but if `(window-start)` is past `(buffer-size)` we'd have a `top-margin` greater than `height`, so `(- height top-margin bottom-margin)` would be negative.

If that's the problem, #445 should fix it. Can you please test if it does, since I'm not sure how to reproduce the problem?

**Comment by seagle0128** [comment]
I can confirm PR #445 fixed this issue.
Actually I can reproduce the issue with this snippet. The key is to display the doc buffer. I guess the current buffer is the doc buffer while displaying the doc. I observed the buffer size is small while not the actual current buffer size. So this segment needs to consider child frame?

```elisp
(use-package company-box
      :diminish
      :hook (company-mode . company-box-mode)
      :init (setq company-box-enable-icon t
                  company-box-backends-colors nil
                  company-box-doc-delay 0.3))
```

My configuration is https://github.com/seagle0128/.emacs.d/blob/64a635a7828de05b4770930d1cc8ec1245a1ae64/lisp/init-company.el#L118.

**Pull request merged**